### PR TITLE
Implement DateTime picker Dialog

### LINKED ISSUE #423: Implement date/time picker
Some Web dialogs are not implemented in Wolvic. These include the color picker and the date/time picker.

After issue https://github.com/Igalia/wolvic/issues/408 has been closed, trying to open one of these dialogs will not crash. However, these dialogs still need to be implemented.

This issue is about implementing the date/time picker that should come up when the user clicks on one of:

* `<input type="date">`
* `<input type="datetime-local">`
* `<input type="month">`
* `<input type="time">`
* `<input type="week">`

### PR DESCRIPTION
**Pull request by HollowMan6**
Implement DateTime picker Dialog

All five types of date/time related inputs are now supported.

Resolves #423

Supersede #640 as rewriting using a more elegant approach and didn't have a base on that.

Screenshot from 2023-07-29 02-10-23
image

### CODE REVIEW

**Comment by HollowMan6** [comment]
See crash when typing in the time picker dialog, not exactly sure what's going on. Suspect that it can actually be some kind of bug related to the Android library. Fixed in: https://github.com/Igalia/wolvic/pull/877/commits/6e95a9bbe9acf8b2162513f9e2b7023891ebf84d

Screenshot from 2023-07-29 13-32-42


```log
handleText              com.igalia.wolvic                    E  8: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
```

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        mPrompt = prompt;
+        initialize(aContext);
+    }
+
+    protected void initialize(Context aContext) {
+        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
+        layoutParams.gravity = Gravity.CENTER_HORIZONTAL;
+        final int padding = 10;
+
+        if (mPrompt.type() == WSession.PromptDelegate.DateTimePrompt.Type.DATE) {
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java:53]: Can we use a switch instead?

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import java.util.Locale;
+
+
+public class DateTimePromptWidget extends PromptWidget {
+    private AudioEngine mAudio;
+    private Button mCancelButton, mClearButton, mOkButton;
+    private WSession.PromptDelegate.DateTimePrompt mPrompt;
+    private String format;
+
+    private static Date parseDate(final SimpleDateFormat formatter, final String value, final boolean defaultToNow) {
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java:32]: Same comment about booleans in APIs. Let's better use a enum and that way we wouldn't need to add clarification comments in the callers

**Review by svillar — CHANGES_REQUESTED** [review]
Great! just a couple of comments

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java (hunk 1 above)
Comment (HollowMan6) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java]: Done!

**Review by svillar — CHANGES_REQUESTED** [review]
This is definitely an improvement over not having it implemented but it requires changes before landing. Basically the widget does not show reliably. The first time you click on the input then the widget is shown, however it isn't possible to show it again most of the times.

**Comment by HollowMan6** [comment]
@svillar  I think when the widget is shown as controlled by the gecko itself, it's not related to the code logic/quality in this PR.

This PR just implements the prompt callback (onDateTimePrompt) and is not in control of when the callback is triggered. For me, I think we are ready to land this PR and we can open another issue for investigating the prompt triggering.

> This is definitely an improvement over not having it implemented but it requires changes before landing. Basically the widget does not show reliably. The first time you click on the input then the widget is shown, however it isn't possible to show it again most of the times.

**Comment by svillar** [comment]
> @svillar I think when the widget is shown as controlled by the gecko itself, it's not related to the code logic/quality in this PR.
> 
> This PR just implements the prompt callback (onDateTimePrompt) and is not in control of when the callback is triggered. For me, I think we are ready to land this PR and we can open another issue for investigating the prompt triggering.
> 
> > This is definitely an improvement over not having it implemented but it requires changes before landing. Basically the widget does not show reliably. The first time you click on the input then the widget is shown, however it isn't possible to show it again most of the times.

Yes Gecko asks us to show it, but we're the ones creating and calling show() on the widget. 

I'd bet that `onDateTimePrompt()` is always properly called. So we might  be doing something wrong there. I haven't actually checked it so if I'm proven wrong then I'd be fine landing this and checking the other issue later.

**Comment by HollowMan6** [comment]
If it's this PR's issue, the only place I can think of that might go wrong is the parameters of this call. https://github.com/Igalia/wolvic/pull/877/files#diff-ab3af07147e220c1dbf95588d7971dd329191e6ae17874b5c30d71598a222bf5R338

I don't have access to the headset now. Can only start debugging this evening.

**Comment by HollowMan6** [comment]
I have started debugging the issue, and it looks like `onDateTimePrompt()` is not properly called (When we finish the prompt and the value is set, onDateTimePrompt() will not get called when we click the input again. The only way is to re-focus the input)

File: app/src/common/shared/com/igalia/wolvic/browser/PromptDelegate.java [hunk 3]
```diff
@@ -317,9 +318,24 @@ public void dismiss() {
     @Nullable
     @Override
     public WResult<PromptResponse> onDateTimePrompt(@NonNull WSession session, @NonNull DateTimePrompt prompt) {
-        // TODO implement date/time picker
         final WResult<PromptResponse> result = WResult.create();
-        result.cancel();
```
Comment (HollowMan6) [inline app/src/common/shared/com/igalia/wolvic/browser/PromptDelegate.java:322]: @svillar Now I believe the issue is caused by `onDateTimePrompt()` not properly called. onDateTimePrompt() will only be called once and will not get called again when we click on the input again. The only way for it to be called again is to re-focus the input.

The issue already exists even without this PR (where we only have the result canceled without showing anything), so it's not about the change of focus on our side.

Comment (svillar) [reply app/src/common/shared/com/igalia/wolvic/browser/PromptDelegate.java:322]: Awesome thanks. Mind opening a new bug with this info? (even just referring to this bug/comment)

**Review by svillar — APPROVED** [review]

File: app/src/common/shared/com/igalia/wolvic/browser/PromptDelegate.java (hunk 3 above)
Comment (HollowMan6) [reply app/src/common/shared/com/igalia/wolvic/browser/PromptDelegate.java:322]: Issue opened at https://github.com/Igalia/wolvic/issues/896

**Pull request merged**